internal/exec/util/file.go

```diff
@@ -33,62 +36,108 @@ const (
 )
 
 type File struct {
-	Path     types.Path
-	Contents []byte
-	Mode     os.FileMode
-	Uid      int
-	Gid      int
+	io.ReadCloser
+	hash.Hash
+	Path types.Path
+	Mode os.FileMode
+	Uid  int
+	Gid  int
+	hash string
 }
 
+func (f File) Verify() error {
+	if f.hash == "" {
+		return nil
+	}
+	sum := f.Sum(nil)
+	encodedSum := make([]byte, hex.EncodedLen(len(sum)))
+	hex.Encode(encodedSum, sum)
+
+	if string(encodedSum) != f.hash {
+		return util.ErrHashMismatch{
+			Calculated: string(encodedSum),
+			Expected:   f.hash,
+		}
+	}
+	return nil
+}
+
+// Returns a *File with a Reader that downloads, sha512's, and unzips the incoming data
+// Returns nil if f had invalid options. Errors reading/verifying/uncompressing the file will
+// present themselves when the Reader is actually read from
 func RenderFile(l *log.Logger, f types.File) *File {
-	var contents []byte
+	var reader io.ReadCloser
 	var err error
+	var hash string
 
-	fetch := func() error {
-		contents, err = util.FetchResource(l, url.URL(f.Contents.Source))
-		return err
+	reader, err = util.FetchResourceAsReader(l, url.URL(f.Contents.Source))
+	if err != nil {
+		l.Crit("Error fetching file %q: %v", f.Path, err)
+		return nil
 	}
 
-	validate := func() error {
-		return util.AssertValid(f.Contents.Verification, contents)
+	hasher, err := util.GetHasher(f.Contents.Verification)
+	if err != nil {
+		l.Crit("Error verifying file %q: %v", f.Path, err)
+		return nil
 	}
 
-	decompress := func() error {
-		contents, err = decompressFile(l, f, contents)
-		return err
+	if hasher != nil {
+		hash = f.Contents.Verification.Hash.Sum
+		reader = struct {
```

Comment:
How about put this in a NewHashedReader function so intent is clear, defining things with anonymous structs inline with real code clutters and distracts a bit.